### ISSUE TITLE
ParamConverterProvider instances are not registered

### ISSUE THREAD

**Issue by mkuchtiak** [oracle/helidon#3796]
## Environment Details
* Helidon Version: 2.4.1
* Helidon SE or Helidon MP: MP
* JDK version: jdk-11.0.9
* OS: Linux Ubuntu
* Docker version (if applicable):

----------

## Problem Description

In the application,  I've registered a custom ParamConverterProvider instances (ExpandParamConverterProvider and InstantParamConverterProvider).

The test is testing ExpandParamConverterProvider (custom implementation of the GET "expand" query parameter)
It happens that neither of these providers are registered for the application.

Note: this is a RANDOM BUG !!!, but i can reproduce it quite often, when running test 

- in IntelliG environment
- in console (**mvn test**)

This issue prevents us to upgrade to Helidon 2.4.1. We are using version 2.3.0 where this issue is not reproducible.

## Steps to reproduce
Run MainTest repeatedly. The test is randomly failing.
(May be you would need to repeat running the test several times to reproduce the issue!!)

Demo application attached.
[quickstart-mp2.zip](https://github.com/oracle/helidon/files/7853059/quickstart-mp2.zip)

**Comment by spericas** [comment]
@mkuchtiak Does this script eventually fail the test for you?

```
while true; do mvn surefire:test; done
```

Been running this for a while and no test failures thus far.

**Comment by spericas** [comment]
I also noticed that you have a self-contained test with its own App and resources. When the test runs, the other apps and resources are also in the mix. Is this all intentional or can we simplify this further by removing classes that are not part of the test itself?

**Comment by mkuchtiak** [comment]
> When the test runs, the other apps and resources are also in the mix. Is this all intentional or can we simplify this further by removing classes that are not part of the test itself?

It's intentional. Seems the error is related to multiple application registration.

Can you try to open the application in IntelliG, and run test from there ?
I think, when you run it from IntelliG probability of failing is higher.

**Comment by spericas** [comment]
@mkuchtiak I was able to reproduce the problem in IntelliJ. Essentially, it is related to the _order in which_ `Application` subclasses are registered and an initialization in Jersey. As you may know, Jersey only supports a _single_ `Application` subclass, and it seems that for `ParamConverterProvider`'s there's an initialization that only takes place once, for the first `Application`  found. 

In your example, if `MainTest$App` is found first, it works, otherwise it fails. You can actually flip the problem around and cause the failure in `Application1` as well (`Application2` is totally irrelevant in this scenario).

At this time, the best solution I can offer based on this Jersey limitation is for you to return _all_ `ParamConverterProvider`'s from _all_ your `Application` subclasses so that they are all properly registered regardless of ordering. Not ideal, but may be our best solution at this time. 

```
public class TestApplication1 extends Application {
    @Override
    public Set<Class<?>> getClasses() {
        return Set.of(InstantParamConverterProvider.class,
                              GreetResource.class, 
                              ExpandParamConverterProvider.class);
    }
}

    @ApplicationPath("/")
    public static class App extends Application {
        @Override
        public Set<Class<?>> getClasses() {
            System.out.println("get classes");
            return ImmutableSet.copyOf(
                    Set.of(
                            ExpandParamConverterProvider.class,
                            InstantParamConverterProvider.class,
                            RootResource.class
                    ));
        }
    }
```

This may have worked before, but it was mostly by accident. Let me know if this acceptable as an initial workaround until a better solution is provided. Alternatively, of course, you could combine all your `Application` subclasses into one.

**Comment by martin-sladecek** [comment]
Hi @spericas, does this mean that when you register a ParamConverterProvider, it may end up being registered for a different Application? Does this affect only ParamConverterProvider or other types are affected as well (like filters, features...).

Unfortunately, the proposed workaround won't work for us. What will happen when we use @Provider or AutoDiscoverable? Will the ParamConverterProvider be available in all applications then?

**Comment by mkuchtiak** [comment]
> Let me know if this acceptable as an initial workaround until a better solution is provided. Alternatively, of course, you could combine all your Application subclasses into one.

This would work for our tests. But, as Martin mentioned, we have no control on the Applications defined in individual microservices,  and the  ParamConverterProvider is defined in a library shared by these microservices.

Is this problem of ParamConverterProvider only or this could happen with any provider specified in Application#getClasses() ?

**Comment by spericas** [comment]
Based on initial analysis of the issue, the problem is specific to how `ParamConverter`'s are initialized during resource registration. I will confirm today, but I don't think we can fix this issue without changes in Jersey.

**Comment by spericas** [comment]
@martin-sladecek @mkuchtiak One thing I forgot to mention is that there's a way to revert back an old (problematic in some cases) behavior where a single injection manager is used for all applications. Just set `server.single-injection-manager=true` in your properties file. This may work for you as well but has some limitations.

**Comment by martin-sladecek** [comment]
@spericas, thanks for the tip. Is there some documentation to help us understand the limitations? I could not find the property in helidon or jersey code.

**Comment by spericas** [comment]
@martin-sladecek No docs on that property yet, but should result in same behavior as pre-2.4.x. Still exploring some alternatives that will not require changes to Jersey ...

**Comment by spericas** [comment]
@martin-sladecek @mkuchtiak There's potentially a way to handle your use case if all `ParamConverterProvder`'s are annotated with `@Provider` --whether returned by `getClasses` or not is irrelevant. That annotation will enable Helidon to find them ahead of time and make them all available to Jersey, as opposed to only a subset depending on which applications are processed first. If this is an acceptable solution in your environment, I can provide a fix in 2.4.2 to be released shortly.

**Comment by spericas** [comment]
PR #3857
PR #3846
